service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,281 @@
+func GetDefaultArguments() Arguments {
+	return Arguments{
+		ID:               agentseed.Get().UID,
+		Metadata:         make(map[string]string),
+		PollFrequency:    1 * time.Minute,
+		HTTPClientConfig: config.CloneDefaultHTTPClientConfig(),
+	}
+}
+
+// SetToDefault implements river.Defaulter.
+func (a *Arguments) SetToDefault() {
+	*a = GetDefaultArguments()
+}
+
+// Validate implements river.Validator.
+func (a *Arguments) Validate() error {
+	// We must explicitly Validate because HTTPClientConfig is squashed and it
+	// won't run otherwise
+	if a.HTTPClientConfig != nil {
+		return a.HTTPClientConfig.Validate()
+	}
+
+	return nil
+}
+
+// Data includes information associated with the remotecfg service.
+type Data struct {
+}
+
+// New returns a new instance of the remotecfg service.
+func New(opts Options) (*Service, error) {
+	basePath := filepath.Join(opts.StoragePath, ServiceName)
+	err := os.MkdirAll(basePath, 0750)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		opts:   opts,
+		ticker: time.NewTicker(math.MaxInt64),
+	}, nil
+}
+
+// Data returns an instance of [Data]. Calls to Data are cachable by the
+// caller.
+//
+// Data must only be called after parsing command-line flags.
+func (s *Service) Data() any {
+	return map[string]string{}
+}
+
+// Definition returns the definition of the remotecfg service.
+func (s *Service) Definition() service.Definition {
+	return service.Definition{
+		Name:       ServiceName,
+		ConfigType: Arguments{},
+		DependsOn:  nil, // remotecfg has no dependencies.
+	}
+}
+
+var _ service.Service = (*Service)(nil)
+
+// Run implements [service.Service] and starts the remotecfg service. It will
+// run until the provided context is canceled or there is a fatal error.
+func (s *Service) Run(ctx context.Context, host service.Host) error {
+	s.ctrl = host.NewController(ServiceName)
+
+	// Run the service's own controller.
+	go func() {
+		s.ctrl.Run(ctx)
+	}()
+
+	// We're on the initial start-up of the service.
+	// Let's try to read from the API, parse and load the response contents.
+	// If either the request itself or parsing its contents fails, try to read
+	// from the on-disk cache on a best-effort basis.
+	var (
+		gcr              *connect.Response[agentv1.GetConfigResponse]
+		src              *flow.Source
+		getErr, parseErr error
+	)
+
+	s.mut.RLock()
+	gcr, getErr = s.asClient.GetConfig(ctx, s.getConfigRequest)
+	if getErr == nil {
+		src, parseErr = flow.ParseSource(ServiceName, []byte(gcr.Msg.GetContent()))
+	}
+
+	// Reading from the API succeeded, let's try to load the contents.
+	if getErr == nil && parseErr == nil {
+		err := s.ctrl.LoadSource(src, nil)
+		if err != nil {
+			level.Error(s.opts.Logger).Log("msg", "could not load the API response contents", "err", err)
+		} else {
+			s.currentConfigHash = getHash(gcr.Msg.Content)
+		}
+	} else {
+		// Either the API call or parsing its contents failed, let's try the
```

Comment:
The risk with retrieving the cached config from the disk is that it may run an old config that I don't want to run anymore just because I misconfigured the connection to my new API or my new config is not correct. In this case I would rather have my agent not run anything than running something I don't want it to run. Maybe there should be an argument to let the user decide what behavior he wants?

Reply:
I'd like to avoid multiple configuration options and try to offer a more opinionated way of people using the service and we've identified this as a necessary feature.

In any case, the hash currently takes into account the entirety of the Arguments block, so it's harder to accidentally reuse an erroneous value.